core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java

```diff
@@ -717,9 +718,10 @@ public RelOptCost getCost(RelNode rel, RelMetadataQuery mq) {
   public RelSubset getSubset(RelNode rel) {
     assert rel != null : "pre: rel != null";
     if (rel instanceof RelSubset) {
-      return (RelSubset) rel;
+      return canonize((RelSubset) rel);
     } else {
-      return mapRel2Subset.get(rel);
+      RelSubset s = mapRel2Subset.get(rel);
+      return s == null ? null : canonize(s);
```

Comment:
IMO, canonize in this method is not needed. It should only be canonized when the subset is created. What do you think?

Reply:
Sorry, wrong button. Not approve.

Reply:
Hmm, is it because the mapping is stale?

Reply:
If the stale subset is required state, during canonize, we may get a subset with delivered. This will lose another info.

Reply:
Good point, fixed